crates/bevy_math/src/primitives/dim3.rs

```diff
@@ -240,12 +240,71 @@ pub struct ConicalFrustum {
 }
 impl Primitive3d for ConicalFrustum {}
 
-/// A torus (AKA donut) primitive.
-#[derive(Clone, Copy, Debug)]
+/// The type of torus determined by the minor and major radii
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum TorusKind {
+    /// A torus that has a ring.
+    /// The major radius is greater than the minor radius
+    Ring,
+    /// A torus that has no hole but also doesn't intersect itself.
+    /// The major radius is equal to the minor radius
+    Horn,
+    /// A self-intersecting torus.
+    /// The major radius is less than the minor radius
+    Spindle,
+}
+
+/// A torus primitive, often representing a ring or donut shape
+#[derive(Clone, Copy, Debug, PartialEq)]
 pub struct Torus {
-    /// The radius of the overall shape
-    pub radius: f32,
-    /// The radius of the internal ring
-    pub ring_radius: f32,
+    /// The radius of the tube of the torus
+    #[doc(
+        alias = "ring_radius",
+        alias = "tube_radius",
+        alias = "cross_section_radius"
+    )]
+    pub minor_radius: f32,
+    /// The distance from the center of the torus to the center of the tube
+    #[doc(alias = "radius_of_revolution")]
+    pub major_radius: f32,
 }
 impl Primitive3d for Torus {}
+
+impl Torus {
+    /// Create a new `Torus` from an inner and outer radius.
+    ///
+    /// The inner radius is the radius of the hole, and the outer radius
+    /// is the radius of the entire object
+    pub fn new(inner_radius: f32, outer_radius: f32) -> Self {
+        let minor_radius = (outer_radius - inner_radius) / 2.0;
+        Self {
+            minor_radius,
+            major_radius: outer_radius - minor_radius,
+        }
+    }
+
+    /// Get the inner radius of the torus.
+    /// For a ring torus, this corresponds to the radius of the hole
+    pub fn inner_radius(&self) -> f32 {
+        self.major_radius - self.minor_radius
+    }
+
+    /// Get the outer radius of the torus.
+    /// This corresponds to the overall radius of the entire object
+    pub fn outer_radius(&self) -> f32 {
+        self.major_radius + self.minor_radius
+    }
+
+    /// Get the [`TorusKind`] determined by the minor and major radii.
+    ///
+    /// The torus can either be a *ring torus* that has a hole,
+    /// a *horn torus* that doesn't have a hole but also isn't self-intersecting,
+    /// or a *spindle torus* that is self-intersecting
+    pub fn kind(&self) -> TorusKind {
+        match self.major_radius.total_cmp(&self.minor_radius) {
```

Comment:
Wouldn't NaN values and +/- 0 values sorta throw this off? NaN should result in an invalid torus and +/- 0 should imo be considered equal.

Maybe this should use `PartialOrd` and return an `Option` or `Result` instead?

Reply:
Agreed about `PartialOrd`, but should I perhaps use `expect` instead and add a "Panics" section in the docs? I would imagine NaN cases to be very rare and the result of otherwise doing something wrong

Reply:
I think panicking is the right choice here.

Reply:
Could this be maybe better by using a comparison with `f32::EPSILON`?
For `NaN` or `+-Inf` values, or negative radii, this method could perhaps return something along the lines of `TorusKind::Invalid` to indicate the torus has some weird non-geometric parameters.

Reply:
An Invalid kind is interesting 🤔

Reply:
I changed it to:
1. panic if the minor or major radius is NaN, and
2. return `TorusKind::Invalid` if the minor or major radius is non-positive or infinite.
Is that fine? Or should it also return `Invalid` for NaN cases? I feel like NaN cases are almost always unwanted and don't really make sense mathematically so a panic might make sense

Reply:
I think return `Invalid` here instead. The user can panic on `Invalid` if they want: we should try to avoid it where possible.

Reply:
Thanks, I changed it to return `Invalid` for `NaN` values as well.